rtloader/common/stringutils.c

```diff
@@ -8,30 +8,29 @@
 #include "rtloader_types.h"
 #include "stringutils.h"
 
-
-PyObject * yload = NULL;
-PyObject * ydump = NULL;
-PyObject * loader = NULL;
-PyObject * dumper = NULL;
+PyObject *yload = NULL;
+PyObject *ydump = NULL;
+PyObject *loader = NULL;
+PyObject *dumper = NULL;
 
 /**
  * returns a C (NULL terminated UTF-8) string from a python string.
  *
  * \param object  A Python string to be converted to C-string.
  *
  * \return A standard C string (NULL terminated character pointer)
- *  The returned pointer is allocated from the heap and must be
- * deallocated (free()ed) by the caller
+ *  The returned pointer is embedded witin PyObject. When string pointer
+ *  is not needed corresponding PyObject reference needs to be decremented
  */
-char *as_string(PyObject *object)
+char *as_embedded_string(PyObject *object, PyObject **stringObject)
```

Comment:
Yes, this is a key piece in the algorithm of having copying Python strings into single destination buffer without their allocation and without "harder" code work on Python buffer string direct access. This is a "trick" to avoid intermedeate allocation while keeping most of the exiting string manipulation code. This is a way to "pin" the python string object and have a safety of direct access to its underlying char* to copy it at will. Only then I deref it. Just look where this function is called from to see it. I consider myself lucky to find this relatively simple approach without major code surgery to achieve desired effect - no intermediate memory allocation and still relatively simple to follow code (IMHO).

Reply:
OK, I see !
Could you add a comment to the function explaining this ?

Reply:
@hush-hush I have added another comment. @truthbk new function allows for outside of stringutils to get length and copy string to other buffer if needed without need to track Python object - it is all done inside stringutils. Old function still exists and used by lot of existing code. Yes, new code is bit riskier but I feel bad keep potentially allocating dozens small intermediate buffer and explicitly freeing across CGO boundary.